### PR TITLE
[BFCL] Handling rate limits

### PR DESCRIPTION
**Pull request by ShishirPatil**
For BFCL generations of hosted models, this will check for rate limit errors and retry thrice. We can choose to include an exponential back-off as is usually recommended, but I think in the interest of keeping it simple this should perhaps be sufficient?

This does NOT affect the leaderboard values.

### CODE REVIEW

**Review by HuanzhiMao — CHANGES_REQUESTED** [review]
1. `retry_count < retry_limit - 1` should be removed, otherwise it would only retry two times (when `retry_count` is 0 and 1).
2. You can also check the status code for indication of rate limit reached. Such as `429 Too Many Requests`, `403 Forbidden` or `503 Service Unavailable`. Purely reply on the text "rate limit reached" might be inaccurate. For example, the below error message from Anthropic API would not be caught:

> Number of requests has exceeded your rate limit. Please try again later or [contact sales](https://www.anthropic.com/contact-sales) to discuss your options for a rate limit increase.

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py [hunk 1]
```diff
@@ -91,7 +91,9 @@ def collect_test_cases(test_filename_total, model_name):
 
 
 def generate_results(args, model_name, test_cases_total):
-    retry_limit = 3
+    RETRY_LIMIT = 3
+    RETRY_DELAY = 65  # Delay in seconds
```
Comment (CharlieJCJ) [inline berkeley-function-call-leaderboard/openfunctions_evaluation.py:95]: Curious where is this number 65 coming from? Add more context here if there are references. 
If this number is arbitrary, let's just put 60 seconds for default 1 minute RETRY_DELAY.

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py [hunk 2]
```diff
@@ -114,9 +117,31 @@ def generate_results(args, model_name, test_cases_total):
...
+                    break  # Success, exit the loop
+                except Exception as e:
+                    # TODO: It might be better to handle the exception in the handler itself rather than a universal catch block here, as each handler use different ways to call the endpoint.
+                    # OpenAI has openai.RateLimitError while Anthropic has anthropic.RateLimitError. It would be more robust in the long run. 
+                    if "rate limit reached" in str(e).lower() or (
+                        hasattr(e, "status_code")
+                        and (
+                            e.status_code == 429
+                            or e.status_code == 503
+                            or e.status_code == 403
```
Comment (CharlieJCJ) [inline berkeley-function-call-leaderboard/openfunctions_evaluation.py:136]: For status code 429, 503, 403 check, I've gathered the following references from openai and anthropic:

#### Anthropic:
403 - permission_error: Your API key does not have permission to use the specified resource.
429 - rate_limit_error: Your account has hit a rate limit.
503 - N/A (no support)
500 - api_error: An unexpected error has occurred internal to Anthropic’s systems.
https://docs.anthropic.com/en/api/errors

#### OpenAI: 
403 - Country, region, or territory not supported
429 - Rate limit reached for requests
429 - You exceeded your current quota, please check your plan and billing details
503 - The engine is currently overloaded, please try again later	
500 - The server had an error while processing your request.
https://platform.openai.com/docs/guides/error-codes/api-errors

Here, 403 is related to permission/auth error, not related to rate limit. 
Only 429 is related to rate limit error. 
500, 503 are related to internal server error and server high traffic, which the official documentation recommend to retry (which can be checked by this PR as well to retry).
Other error codes in the documentations (e.g. 401) are not very related. 

**Suggest change**: 
- Remove e.status_code == 403 check
- Add e.status_code == 500 check.

**Review by CharlieJCJ — CHANGES_REQUESTED** [review]

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py (hunk 1 above)
Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/openfunctions_evaluation.py:95]: 60s for the timer to complete. But often we find that even with 60 there is a conflict. So 65 is a safe nos

Comment (CharlieJCJ) [reply berkeley-function-call-leaderboard/openfunctions_evaluation.py:95]: Ok sounds good, can you add this as a comment in code for reference?

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py (hunk 2 above)
Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/openfunctions_evaluation.py]: changed

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py (hunk 1 above)
Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/openfunctions_evaluation.py:95]: sure

**Review by HuanzhiMao — APPROVED** [review]

**Review by CharlieJCJ — APPROVED** [review]

**Pull request merged**